[flutter_tools] ensure track-widget-creation can be changed on devcompiler

2ZeroSix opened the pull request:
## Description

Workaround for cached web builds with `--[no-]track-widget-creation` flags.
Similar PR: #23299

## Related Issues

closes #56212

## Tests

I added the following tests:
enable devicelab web_enable_twc

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

2ZeroSix committed:
pass track widget creation

2ZeroSix committed:
wrap cache kernel path with getKernelPathForTransformerOptions

2ZeroSix committed:
enable web_enable_twc test

2ZeroSix committed:
fix indentation

jonahwilliams commented:
This approach looks good to me. Please LMK when you're ready for code review, or if you want any advice on testing

jonahwilliams commented:
Most of the failures are caused by this line of mocking:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/resident_web_runner_test.dart#L144

If you update that to be `anyNamed('trackWidgetCreation'` it should just work

2ZeroSix committed:
update trackWidgetCreation in mocks

`trackWidgetCreation: true` =>  `trackWidgetCreation: anyNamed('trackWidgetCreation')`

2ZeroSix committed:
get buildInfo from debuggingOptions instead of flutterDevices

2ZeroSix commented:
@jonahwilliams 
> Most of the failures are caused by this line of mocking:
> 
> https://github.com/flutter/flutter/blob/master/packages/flutter_tools/test/general.shard/resident_web_runner_test.dart#L144
> 
> If you update that to be `anyNamed('trackWidgetCreation'` it should just work

Thanks! I hope this change in tests shouldn't be considered as breaking change, since it seems to be expected behavior?

> This approach looks good to me. Please LMK when you're ready for code review, or if you want any advice on testing

Finally all tests should be green. Can you review it?

jonahwilliams commented on packages/flutter_tools/lib/src/resident_runner.dart:
```diff
@@ -62,6 +63,14 @@ class FlutterDevice {
        );
 
+
+  static String getCachedDillPath({@required bool trackWidgetCreation}) {
```
nit: I would make this a top level instead of a static, since it isn't really related to the FlutterDevice

2ZeroSix replied:
moved to `bundle.dart` as top level `getDefaultCachedKernelPath`

jonahwilliams commented on packages/flutter_tools/lib/src/resident_runner.dart:
```diff
@@ -1179,10 +1193,11 @@ abstract class ResidentRunner {
     // If _dillOutputPath is null, we created a temporary directory for the dill.
     if (_dillOutputPath == null && artifactDirectory.existsSync()) {
-      final File outputDill = artifactDirectory.childFile('app.dill');
+      final File outputDill = globals.fs.file(dillOutputPath);
```
Why did this line change? `_dillOutputPath` could be null here

2ZeroSix replied:
Yep, but here's how `dillOutputPath` defined
```dart
String get dillOutputPath => _dillOutputPath ?? globals.fs.path.join(artifactDirectory.path, 'app.dill');
```

There were only 2 places where `app.diil` was hard-coded and I thought it was a good idea to reduce this to 1

jonahwilliams replied:
Ahh makes sense, thanks!

2ZeroSix committed:
redefine FlutterDevice.getCachedDillPath as getDefaultCachedKernelPath

jonahwilliams reviewed (approved):
LGTM

jonahwilliams commented:
Thank you for fixing this @2ZeroSix

jonahwilliams commented:
Will manually land once CI finished
